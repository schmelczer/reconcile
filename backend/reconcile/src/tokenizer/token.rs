#[derive(Debug, Clone)]
pub struct Token<T>
where
    T: PartialEq + Clone,
{
    normalised: T,
    original: String,
}

impl From<&str> for Token<String> {
    fn from(s: &str) -> Self {
        Token {
            normalised: s.to_string(),
            original: s.to_string(),
        }
    }
}

impl<T> Token<T>
where
    T: PartialEq + Clone,
{
    pub fn new(normalised: T, original: String) -> Self {
        Token {
            normalised,
            original,
        }
    }

    pub fn original(&self) -> &str {
        &self.original
    }

    pub fn normalised(&self) -> &T {
        &self.normalised
    }

    pub fn get_original_length(&self) -> usize {
        self.original.chars().count()
    }
}

impl<T> PartialEq for Token<T>
where
    T: PartialEq + Clone,
{
    fn eq(&self, other: &Self) -> bool {
        self.normalised == other.normalised
    }
}
