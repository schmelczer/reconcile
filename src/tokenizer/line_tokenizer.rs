use super::token::Token;

/// Splits text into lines, preserving line endings as separate tokens.
///
/// ## Example
///
/// ```not_rust
/// "Hello\nWorld!" -> ["Hello", "\n", "World!"]
/// "Line 1\r\nLine 2" -> ["Line 1", "\r\n", "Line 2"]
/// ```
pub fn line_tokenizer(text: &str) -> Vec<Token<String>> {
    let mut result = Vec::new();
    let mut line_start = 0;
    
    let mut chars = text.char_indices().peekable();
    while let Some((i, c)) = chars.next() {
        if c == '\n' {
            // Add line content if any
            if i > line_start {
                result.push(text[line_start..i].into());
            }
            // Add newline
            result.push("\n".into());
            line_start = i + 1;
        } else if c == '\r' && chars.peek() == Some(&(i + 1, '\n')) {
            // Handle \r\n
            if i > line_start {
                result.push(text[line_start..i].into());
            }
            chars.next(); // consume \n
            result.push("\r\n".into());
            line_start = i + 2;
        }
    }
    
    // Add final line if any
    if line_start < text.len() {
        result.push(text[line_start..].into());
    }
    
    result
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn test_with_snapshots() {
        assert_debug_snapshot!(line_tokenizer(""));

        assert_debug_snapshot!(line_tokenizer("Hello"));

        assert_debug_snapshot!(line_tokenizer("Hello\nWorld"));

        assert_debug_snapshot!(line_tokenizer("Hello\nWorld\n"));

        assert_debug_snapshot!(line_tokenizer("Line 1\r\nLine 2"));

        assert_debug_snapshot!(line_tokenizer("Multi\nLine\nText\nHere"));

        assert_debug_snapshot!(line_tokenizer("\n"));

        assert_debug_snapshot!(line_tokenizer("\n\n"));

        assert_debug_snapshot!(line_tokenizer("Start\n\nEnd"));
    }
}
