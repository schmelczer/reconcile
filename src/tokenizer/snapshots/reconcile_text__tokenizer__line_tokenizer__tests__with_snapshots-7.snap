---
source: src/tokenizer/line_tokenizer.rs
expression: "line_tokenizer(\"\\n\")"
snapshot_kind: text
---
[
    Token {
        normalized: "\n",
        original: "\n",
        is_left_joinable: true,
        is_right_joinable: true,
    },
]
